### PR TITLE
Apply pod name and namespace labels for pod cgroup for cadvisor metrics

### PR DESCRIPTION
**Pull request by derekwaynecarr**
**What this PR does / why we need it**:
1. Enable Prometheus users to determine usage by pod name and namespace for pod cgroup sandbox.
1. Label cAdvisor metrics for pod cgroups by pod name and namespace.
1. Aligns with kubelet stats summary endpoint pod cpu and memory stats.

**Special notes for your reviewer**:
This provides parity with the summary API enhancements done here:
https://github.com/kubernetes/kubernetes/pull/55969

**Release note**:
```release-note
Apply pod name and namespace labels to pod cgroup in cAdvisor metrics
```

### CODE REVIEW

**Comment by derekwaynecarr** [comment]
/hold

want to collect feedback, but i think this is useful to understand via prometheus usage of pod sandbox by pod name and namespace.

/cc @sjenning @smarterclayton @dashpole @brancz @DirectXMan12

**Comment by derekwaynecarr** [comment]
ok, this should be ready for review.

File: pkg/kubelet/cm/pod_container_manager_linux.go [hunk 1]
```diff
@@ -184,6 +184,27 @@ func (m *podContainerManagerImpl) ReduceCPULimits(podCgroup CgroupName) error {
 	return m.cgroupManager.ReduceCPULimits(podCgroup)
 }
 
+// IsPodCgroup returns true if the cgroup name corresponds to a pod
+func (m *podContainerManagerImpl) IsPodCgroup(podCgroup CgroupName) (bool, types.UID) {
```
Comment (dashpole) [inline pkg/kubelet/cm/pod_container_manager_linux.go]: I would love to see a unit test for this

**Comment by dashpole** [comment]
I like this change.  I need to take a closer look, but it would be nice to make this and #55969 use the same logic for identifying pods so we dont get inconsistent results or bugs in one, but not the other.

**Comment by dashpole** [comment]
@derekwaynecarr agreed, #55969 takes the approach of iterating over all pods and doing `Pod API Object -> Pod Cgroup  Name -> Query Pod Cgroup Name in cAdvisor`.  This does the reverse, as we iterate over all cgroups, and determine if each one belongs to a pod by extracting the UID and querying for it in our known pods.  So we can't share any of the logic.

File: pkg/kubelet/cm/pod_container_manager_linux.go [hunk 2]
```diff
@@ -184,6 +184,31 @@ func (m *podContainerManagerImpl) ReduceCPULimits(podCgroup CgroupName) error {
 	return m.cgroupManager.ReduceCPULimits(podCgroup)
 }
 
+// IsPodCgroup returns true if the literal cgroupfs name corresponds to a pod
+func (m *podContainerManagerImpl) IsPodCgroup(cgroupfs string) (bool, types.UID) {
+	// convert the literal cgroupfs form to the driver specific value
+	cgroupName := m.cgroupManager.CgroupName(cgroupfs)
+	qosContainersList := [3]CgroupName{m.qosContainersInfo.BestEffort, m.qosContainersInfo.Burstable, m.qosContainersInfo.Guaranteed}
+	basePath := ""
+	for _, qosContainersList := range qosContainersList {
```
Comment (dashpole) [inline pkg/kubelet/cm/pod_container_manager_linux.go]: nit: rename either of the qosContainersList variables to avoid confusion.

**Review by dashpole — COMMENTED** [review]
one nit, otherwise LGTM.  Waiting to add lgtm for question about prometheus labels to be resolved.

**Comment by derekwaynecarr** [comment]
@dashpole updated, thx.

**Comment by DirectXMan12** [comment]
Let me respond to that one when I'm not still in post-KubeCon haze (hopefully tomorrow -- I think I caught some con crud over the weekend).

File: pkg/kubelet/server/server.go [hunk 3]
```diff
@@ -825,31 +826,41 @@ func (a prometheusHostAdapter) GetMachineInfo() (*cadvisorapi.MachineInfo, error
...
+			if pod, found := s.GetPodByCgroupName(cgroupName); found {
+				podName = pod.Name
+				namespace = pod.Namespace
+			}
+		}
+		set := map[string]string{
+			metrics.LabelID:    c.Name,
+			metrics.LabelName:  name,
+			metrics.LabelImage: image,
+			"pod_name":         podName,
```
[the remark this answers is not in the document]
Comment (derekwaynecarr) [reply pkg/kubelet/server/server.go]: this just keeps the existing naming extended to pod cgroups.   We can change or discuss label names in a follow-on

[the remark this answers is not in the document]
Comment (DirectXMan12) [reply pkg/kubelet/server/server.go]: yeah, I'd like to change it at some point, but let's keep it as is in this PR.

File: pkg/kubelet/cm/pod_container_manager_linux.go [hunk 4]
```diff
@@ -184,6 +184,31 @@ func (m *podContainerManagerImpl) ReduceCPULimits(podCgroup CgroupName) error {
 	return m.cgroupManager.ReduceCPULimits(podCgroup)
 }
 
+// IsPodCgroup returns true if the literal cgroupfs name corresponds to a pod
+func (m *podContainerManagerImpl) IsPodCgroup(cgroupfs string) (bool, types.UID) {
+	// convert the literal cgroupfs form to the driver specific value
+	cgroupName := m.cgroupManager.CgroupName(cgroupfs)
+	qosContainersList := [3]CgroupName{m.qosContainersInfo.BestEffort, m.qosContainersInfo.Burstable, m.qosContainersInfo.Guaranteed}
```
Comment (DirectXMan12) [inline pkg/kubelet/cm/pod_container_manager_linux.go]: Is this list used elsewhere?  If so, can we move it up to a constant so we have a single source of truth?

File: pkg/kubelet/cm/pod_container_manager_linux.go [hunk 5]
```diff
@@ -184,6 +184,31 @@ func (m *podContainerManagerImpl) ReduceCPULimits(podCgroup CgroupName) error {
...
+			basePath = cgroupName[len(qosContainerName)]
+		}
+	}
+	if basePath == "" {
+		return false, types.UID("")
+	}
+	if !strings.HasPrefix(basePath, podCgroupNamePrefix) {
+		return false, types.UID("")
+	}
+	parts := strings.Split(basePath, podCgroupNamePrefix)
```
Comment (DirectXMan12) [inline pkg/kubelet/cm/pod_container_manager_linux.go]: this looks weird.  if it's just a prefix, `basePath[len(podCgroupNamePrefix):]` is clearer, IMO

**Review by DirectXMan12 — APPROVED** [review]
couple nits inline, otherwise looks good to me

File: pkg/kubelet/server/server.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (brancz) [reply pkg/kubelet/server/server.go]: sounds reasonable to me

**Comment by sjenning** [comment]
I'm going to go ahead and
/lgtm
nits can be fixed in follow on since there doesn't be any fundamental for functional objection
/retest

**Comment by serathius** [comment]
Looks like this PR is cause of failing tests https://k8s-testgrid.appspot.com/sig-instrumentation#gce-prometheus
It introduced additional timeseries per pod causing cpu queries return doubled result.
Was this an intended behaviour? should we introduce breaking change in queries and require users to fix all their dashboard by adding additional filter (e. g. `image!=""` or pattern match on `id`)

**Comment by brancz** [comment]
I'm not sure we are ready to provide any compatibility guarantees around metrics at this point, especially with cadvisor being an external project. In the past we have broken metrics several times simply because cadvisor had changes. I think we can only confidently provide stability around metrics with the new node-metrics proposal that's currently being discussed in sig-instrumentation.

**Comment by serathius** [comment]
thanks for explaining.
I need to catch up with proposals

**Pull request merged**